### ISSUE TITLE
降低显存的方法 (供参考)

### ISSUE THREAD

**Issue by peizhiluo007** [TemryL/ComfyUI-IDM-VTON#45]
在下面函数(ComfyUI-IDM-VTON\src\nodes\pipeline_loader.py)，
def load_pipeline(self, weight_dtype):

修改两点：
修改1> 把所有的模型加后的 .to(DEVICE) 全部注释掉。 
修改2> 110行处
        修改前:
               pipe.unet_encoder = unet_encoder
               pipe = pipe.to(DEVICE)
               pipe.weight_dtype = weight_dtype
        修改前:
               pipe.enable_sequential_cpu_offload()
               pipe.unet_encoder.to(DEVICE)
               #pipe.to(DEVICE)

在显卡12G测试，完全无压力。查看显存占用大概6G多点，估计在8G下也能跑。

**Comment by peizhiluo007** [comment]
class PipelineLoader:
    @classmethod
    def INPUT_TYPES(s):
        return {
            "required": {
                "weight_dtype": (("float32", "float16", "bfloat16"), ),
            }
        }
    
    CATEGORY = "ComfyUI-IDM-VTON"
    INPUT_NODE = True
    RETURN_TYPES = ("PIPELINE",)
    FUNCTION = "load_pipeline"
    
    def load_pipeline(self, weight_dtype):
        if weight_dtype == "float32":
            weight_dtype = torch.float32
        elif weight_dtype == "float16":
            weight_dtype = torch.float16
        elif weight_dtype == "bfloat16":
            weight_dtype = torch.bfloat16
        print("[load]----------->noise_scheduler")
        noise_scheduler = DDPMScheduler.from_pretrained(
            WEIGHTS_PATH, 
            subfolder="scheduler"
        )
        print("[load]----------->vae")
        vae = AutoencoderKL.from_pretrained(
            WEIGHTS_PATH,
            subfolder="vae",
            torch_dtype=weight_dtype
        ).requires_grad_(False).eval()#.to(DEVICE)
        print("[load]----------->unet")
        unet = UNet2DConditionModel.from_pretrained(
            WEIGHTS_PATH,
            subfolder="unet",
            torch_dtype=weight_dtype
        ).requires_grad_(False).eval()#.to(DEVICE)
        print("[load]----------->image_encoder")
        image_encoder = CLIPVisionModelWithProjection.from_pretrained(
            WEIGHTS_PATH,
            subfolder="image_encoder",
            torch_dtype=weight_dtype
        ).requires_grad_(False).eval()#.to(DEVICE)
        print("[load]----------->unet_encoder")
        unet_encoder = UNet2DConditionModel_ref.from_pretrained(
            WEIGHTS_PATH,
            subfolder="unet_encoder",
            torch_dtype=weight_dtype
        ).requires_grad_(False).eval()#.to(DEVICE)
        print("[load]----------->text_encoder_one")
        text_encoder_one = CLIPTextModel.from_pretrained(
            WEIGHTS_PATH,
            subfolder="text_encoder",
            torch_dtype=weight_dtype
        ).requires_grad_(False).eval()#.to(DEVICE)
        print("[load]----------->text_encoder_two")
        text_encoder_two = CLIPTextModelWithProjection.from_pretrained(
            WEIGHTS_PATH,
            subfolder="text_encoder_2",
            torch_dtype=weight_dtype
        ).requires_grad_(False).eval()#.to(DEVICE)
        print("[load]----------->tokenizer_one")
        tokenizer_one = AutoTokenizer.from_pretrained(
            WEIGHTS_PATH,
            subfolder="tokenizer",
            revision=None,
            use_fast=False,
        )
        print("[load]----------->tokenizer_two")
        tokenizer_two = AutoTokenizer.from_pretrained(
            WEIGHTS_PATH,
            subfolder="tokenizer_2",
            revision=None,
            use_fast=False,
        )
        print("[load]----------->build TryonPipeline")
        pipe = TryonPipeline.from_pretrained(
            WEIGHTS_PATH,
            unet=unet,
            vae=vae,
            feature_extractor=CLIPImageProcessor(),
            text_encoder=text_encoder_one,
            text_encoder_2=text_encoder_two,
            tokenizer=tokenizer_one,
            tokenizer_2=tokenizer_two,
            scheduler=noise_scheduler,
            image_encoder=image_encoder,
            torch_dtype=weight_dtype,
        )
        print("[load]----------->attach unet_encoder")
        pipe.weight_dtype = weight_dtype
        pipe.unet_encoder = unet_encoder
        pipe.enable_sequential_cpu_offload()
        pipe.unet_encoder.to(DEVICE)
        #pipe.to(DEVICE)
        #
        return (pipe, )

**Comment by TemryL** [comment]
Wow that's awesome! Thanks! Could you open a PR with these changes?

**Comment by lldacing** [comment]
速度会变慢吗

**Comment by qtmssa** [comment]
Does this work?

**Comment by qtmssa** [comment]
My 4090 (with 24G vRAM) still OOM, :-(
Anybody help? :-)

**Comment by deepfree2023** [comment]
Works without problem.

**Comment by peizhiluo007** [comment]
> Wow that's awesome! Thanks! Could you open a PR with these changes?

OK

**Comment by dachangqing** [comment]
大佬，为什么我照你说的改了代码，还是出现内存不足的报错啊

**Comment by 925-Studio** [comment]
Thanks for sharing this tip, it works fine.

**Comment by zmwv823** [comment]
![image](https://github.com/TemryL/ComfyUI-IDM-VTON/assets/13308350/51208283-0e31-48d6-b44f-fea32c732c52)
升级了diffusers版本到0.29.2，发现这个插件挂了，然后自己试着改了下，又发现你这有个时间换硬件的办法，就测试玩一下(4G显存太小，懒得跑)。
测试了确实可行，我这分阶段生成，先生成遮罩图，然后再跑换装。
换装过程显存占用大概在6G左右，最后解码飙到7G左右。
12显存的话，你可以试着挂一部分模型到显卡(可能会出现不在同一设备的错误)，提升速度。
![image](https://github.com/TemryL/ComfyUI-IDM-VTON/assets/13308350/cbda4296-54ee-4e54-9971-f4bd1d6ec7d1)

**Comment by TemryL** [comment]
Wow awesome, thank you so much for this finding! Could you create a PR for this?

**Comment by zmwv823** [comment]
**Schedulers from diffusers:**
```
......
                "scheduler": (scheduler_list,),
                "scheduler_karras" :("BOOLEAN", {"default": False}),
......

......
        if scheduler == "DDPM_From_Model":
            noise_scheduler = DDPMScheduler.from_pretrained(
                WEIGHTS_PATH, 
                subfolder="scheduler"
            )
        else:
            noise_scheduler = get_sdxl_sheduler(scheduler, scheduler_karras)
......
scheduler_list = [
    "Euler a",
    "DPM++ 2M",
    "DPM++ 2M SDE",
    "DPM++ SDE",
    "DPM2",
    "DPM2 a",
    "Euler",
    "DDIM",
    "Heun",
    "LMS",
    "UniPC",
    "TCD",
    "LCM",
    "deis",
    "ipndm",
    "DDPM",
    "DDPM_From_Model",
]
```
```
from diffusers import DPMSolverMultistepScheduler, DPMSolverSinglestepScheduler, KDPM2DiscreteScheduler, KDPM2AncestralDiscreteScheduler, EulerDiscreteScheduler, EulerAncestralDiscreteScheduler, HeunDiscreteScheduler, LMSDiscreteScheduler, UniPCMultistepScheduler, DDIMScheduler, TCDScheduler, LCMScheduler, IPNDMScheduler, DEISMultistepScheduler

def get_sdxl_sheduler(name, use_karras):
# sdxl schedulers, sd1.5 schedulers is different.
    scheduler = False
    if name == "DPM++ 2M":
        scheduler = DPMSolverMultistepScheduler(use_karras_sigmas=use_karras)
    elif name == "DPM++ 2M SDE":
        scheduler = DPMSolverMultistepScheduler(algorithm_type="sde-dpmsolver++", use_karras_sigmas=use_karras)
    elif name == "DPM++ SDE":
        scheduler = DPMSolverSinglestepScheduler(use_karras_sigmas=use_karras) 
    elif name == "DPM2":
        scheduler = KDPM2DiscreteScheduler(use_karras_sigmas=use_karras) 
    elif name == "DPM2 a":
        scheduler = KDPM2AncestralDiscreteScheduler(use_karras_sigmas=use_karras) 
    elif name == "Euler":
        scheduler = EulerDiscreteScheduler(use_karras_sigmas=use_karras) 
    elif name == "Euler a":
        scheduler = EulerAncestralDiscreteScheduler() 
    elif name == "Heun":
        scheduler = HeunDiscreteScheduler(use_karras_sigmas=use_karras) 
    elif name == "LMS":
        scheduler = LMSDiscreteScheduler(use_karras_sigmas=use_karras) 
    elif name == "UniPC":
        scheduler = UniPCMultistepScheduler(use_karras_sigmas=use_karras) 
    elif name == "DDIM":
        scheduler = DDIMScheduler() 
    elif name == "TCD":
        scheduler = TCDScheduler() 
    elif name == "LCM":
        scheduler = LCMScheduler() 
    elif name == "deis":
        scheduler = DEISMultistepScheduler(use_karras_sigmas=use_karras)
    elif name == "ipndm":
        scheduler = IPNDMScheduler()
    elif name == "DDPM":
        scheduler = DDPMScheduler()
    return scheduler
```

**Comment by Jeff-goal** [comment]
升级以后这个方法报错，vton无法导入了，请问该如何修改？

**Comment by zhucenichenghao** [comment]
it works，老铁

**Comment by lindkkk** [comment]
大佬太牛逼了，我搞了一周，每次都是爆显存，我还是双4070tis 16G显卡，没想到这么简单就解决了。提醒下修改后找不到idm-vton节点的兄弟们，换的那几行缩进 不能用tab键，要用空格，我被这个问题也卡了1小时
